x/gateway/keeper/msg_server_stake_gateway.go

```diff
@@ -24,46 +26,72 @@ func (k msgServer) StakeGateway(
+	// NB: This SHOULD NEVER happen because msg.ValidateBasic() validates the address as bech32.
+	if err != nil {
+		// TODO_TECHDEBT(#384): determine whether to continue using cosmos logger for debug level.
+		logger.Info(fmt.Sprintf("could not parse address %q", msg.Address))
+		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 
 	// Check if the gateway already exists or not
-	var err error
-	var coinsToEscrow sdk.Coin
+	var (
+		coinsToEscrow sdk.Coin
+		totalStake    sdk.Coin
+	)
 	gateway, isGatewayFound := k.GetGateway(ctx, msg.Address)
 	if !isGatewayFound {
-		logger.Info(fmt.Sprintf("Gateway not found. Creating new gateway for address %q", msg.Address))
+		logger.Info(fmt.Sprintf("gateway not found. Creating new gateway for address %q", msg.Address))
 		gateway = k.createGateway(ctx, msg)
 		coinsToEscrow = *msg.Stake
+		totalStake = *msg.Stake
 	} else {
-		logger.Info(fmt.Sprintf("Gateway found. About to try and update gateway for address %q", msg.Address))
+		logger.Info(fmt.Sprintf("gateway found. About to try and update gateway for address %q", msg.Address))
 		currGatewayStake := *gateway.Stake
 		if err = k.updateGateway(ctx, &gateway, msg); err != nil {
 			logger.Error(fmt.Sprintf("could not update gateway for address %q due to error %v", msg.Address, err))
-			return nil, err
+			return nil, status.Error(codes.InvalidArgument, err.Error())
 		}
 		coinsToEscrow, err = (*msg.Stake).SafeSub(currGatewayStake)
 		if err != nil {
-			return nil, err
+			return nil, status.Error(
+				codes.InvalidArgument,
+				types.ErrGatewayInvalidStake.Wrapf(
+					"stake (%s) must be higher than previous stake (%s)",
+					msg.Stake, currGatewayStake,
+				).Error(),
+			)
 		}
-		logger.Info(fmt.Sprintf("Gateway is going to escrow an additional %+v coins", coinsToEscrow))
+		totalStake = currGatewayStake.Add(*msg.Stake)
+		logger.Info(fmt.Sprintf("gateway is going to escrow an additional %+v coins", coinsToEscrow))
 	}
 
-	// Must always stake or upstake (> 0 delta)
+	// MUST ALWAYS stake or upstake (> 0 delta).
 	if coinsToEscrow.IsZero() {
-		logger.Warn(fmt.Sprintf("Gateway %q must escrow more than 0 additional coins", msg.Address))
-		return nil, types.ErrGatewayInvalidStake.Wrapf("gateway %q must escrow more than 0 additional coins", msg.Address)
+		errFmt := "gateway %q must escrow more than 0 additional coins"
+		logger.Info(fmt.Sprintf(errFmt, msg.Address))
+		return nil, status.Error(
+			codes.InvalidArgument,
+			types.ErrGatewayInvalidStake.Wrapf(errFmt, msg.Address).Error(),
+		)
 	}
 
-	// Retrieve the address of the gateway
-	gatewayAddress, err := sdk.AccAddressFromBech32(msg.Address)
-	if err != nil {
-		// TODO_TECHDEBT(#384): determine whether to continue using cosmos logger for debug level.
-		logger.Error(fmt.Sprintf("could not parse address %q", msg.Address))
-		return nil, err
+	// MUST ALWAYS have at least minimum stake.
+	minStake := k.GetParams(ctx).MinStake
+	if totalStake.Amount.LT(minStake.Amount) {
```

Comment:
As per my previous comment, this should be:
```suggestion
	if msg.Stake.Amount.LT(minStake.Amount) {
```